add logic for BlockOperators in simplify

ullmannsven opened the pull request:
This PR add support for BlockOperators in algorithms/simplify.py. 

These changes allow for a decomposition of BlockOperators, that contain Lincomb or Adjoint operators into LincombOperators of BlockOperators, with parameter dependency removed from Block. This is required for an efficient offline-online decomposition of BlockOperators, as introduced in #2448.

sdrave reviewed (changes requested):
Thanks, looks very good already. I only found one real issue (the `replace_children` call).

Sorry for not having said this before, but could you break this up even more into three PRs (source_spaces/range_spaces, conjugate, simplify). I think it's really three separate features (or four, if you let BlockOperators have children).

Apart from my code comments, could also:

- extend the documentation of `expand` to mention Blockoperators
- write a few simple tests for the new features?

sdrave commented:
@ullmannsven, can you rebase on main to get a clean diff?

ullmannsven commented:
> @ullmannsven, can you rebase on main to get a clean diff?

Done. I still need to write tests, extend documentation and include the handling for the ConcatenationOperators now.

sdrave commented:
pre-commit.ci autofix

ullmannsven commented on src/pymor/algorithms/simplify.py:
```diff
@@ -97,6 +120,73 @@ def action_LincombOperator(self, op):
+                else:
+                    cur = const_blocks[i, j]
+                    if cur is None:
+                        const_blocks[i, j] = contract(c_k * a_k)
```
i thought about my implementation for a bit and I am not sure anymore if calling `contract` at this point makes any sense. Currently, `èxpand` gets called on the `a_k`, that is already one operator of the `LincombOperator` `b`. If for example there were nested `LincombOperators` in one block of the the initial `BlockOperator` that should be expanded, than the `replace_children` call handles this nestedness. (same for nested `ConcatenationOperators`). Then i dont see what `contract` would do at this point, as it will never get called on a `LincombOperator` or a `ConcatenationOperator`. I think the `contract` call needs to be done after the inner loop and then look like  `const_blocks[i, j] = contract(const_block[i, j])`. This would contract the `LincombOperator` that gets created with the `const_blocks[i, j] += c_k * a_k` in line 176. And then there we could move this into a non-const block if this `contract` call fails and returns a `LincombOperator`.

ullmannsven commented:
As discussed, i completely removed the `contract` method from `expand`, in order to not create confusion and have most flexibility for the user. I added a test `test_block_expand_and_contract_only_constant_blocks()` that calls first `expand` and then `contract`, which shows that desired behavior that was before done only by the `expand` call. As `contract` is not part of of `expand` anymore, i changed the docstring again. I also added a tests for the behavior of an `AdjointOperator` of a `ConcatentationOperator` of length one (shoud remove the concatenation). Thinking about this, this is again something that is also done by `contract`. However,  i feel like this edge case is a reasonable simplification done by `expand` and for me its okay to leave it like this. Therefore, from my end, this would be ready to merge.

sdrave commented on src/pymor/algorithms/simplify.py:
```diff
@@ -97,6 +118,75 @@ def action_LincombOperator(self, op):
+                    if cur is None:
+                        const_blocks[i, j] = c_k * a_k
+                    else:
+                        const_blocks[i, j] += c_k * a_k
```
```suggestion
                blocks_k = np.full((nrows, ncols), None, dtype=object)
                blocks_k[i, j] = a_k
                expanded_terms.append((c_k, BlockOperator(blocks_k, range_spaces=op.range.subspaces,
                                                        source_spaces=op.source.subspaces)))
```

Maybe I'm confused, but isn't this more in line with that we have discussed? The else clause would cause `const_part` to have `LincombOperators` as blocks, but the main purpose of `expand` is to move `LincombOperators` from the inside to the outside? When `contract` is called afterwards, both version should give the same result.
